### PR TITLE
[RISCV] Prevent RISCVMergeBaseOffsetOpt from calling getVRegDef on a physical register.

### LINKED ISSUE #78679: RISCV64 backend segfault in RISC-V Merge Base Offset
This program fails with -O3
```/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang red.c -O3```

Discovered/tested using 430a40d12eaa5a61792c4670955c110146902afb (not bisected)

https://godbolt.org/z/zrxq4fdME

Testcase
```c
int printf(const char *, ...);
int b, c, d;
long e, f;
char g[100][100];
long **h;
static int i;
int k, l;
static int *m = &k;
int main() {
  int n = 2;
  int **o = &m;
  int ***p = &o;
  for (c = 0; c >= 0;) {
    int *q = &n;
    *m = 0;
    **h = 0 == p;
    for (l = 1; l < 5; l++)
      if (g[l][c + 3])
        break;
    f = i--;
    e = n;
    *q = 0;
  }
  for (int j = 0;; j++) {
    b = g[d][j];
    printf("Hello\n");
  }
}
```

<details>
  <summary>Crash log (click)</summary>
  
  ```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18 -cc1 -triple riscv64-unknown-linux-gnu -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name red.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu generic-rv64 -target-feature +m -target-feature +a -target-feature +f -target-feature +d -target-feature +c -target-feature +zicsr -target-feature -e -target-feature -h -target-feature -smaia -target-feature -ssaia -target-feature -svinval -target-feature -svnapot -target-feature -svpbmt -target-feature -v -target-feature -xcvalu -target-feature -xcvbi -target-feature -xcvbitmanip -target-feature -xcvelw -target-feature -xcvmac -target-feature -xcvmem -target-feature -xcvsimd -target-feature -xsfvcp -target-feature -xsfvfnrclipxfqf -target-feature -xsfvfwmaccqqq -target-feature -xsfvqmaccdod -target-feature -xsfvqmaccqoq -target-feature -xtheadba -target-feature -xtheadbb -target-feature -xtheadbs -target-feature -xtheadcmo -target-feature -xtheadcondmov -target-feature -xtheadfmemidx -target-feature -xtheadmac -target-feature -xtheadmemidx -target-feature -xtheadmempair -target-feature -xtheadsync -target-feature -xtheadvdot -target-feature -xventanacondops -target-feature -zawrs -target-feature -zba -target-feature -zbb -target-feature -zbc -target-feature -zbkb -target-feature -zbkc -target-feature -zbkx -target-feature -zbs -target-feature -zca -target-feature -zcb -target-feature -zcd -target-feature -zce -target-feature -zcf -target-feature -zcmp -target-feature -zcmt -target-feature -zdinx -target-feature -zfa -target-feature -zfh -target-feature -zfhmin -target-feature -zfinx -target-feature -zhinx -target-feature -zhinxmin -target-feature -zicbom -target-feature -zicbop -target-feature -zicboz -target-feature -zicntr -target-feature -zifencei -target-feature -zihintntl -target-feature -zihintpause -target-feature -zihpm -target-feature -zk -target-feature -zkn -target-feature -zknd -target-feature -zkne -target-feature -zknh -target-feature -zkr -target-feature -zks -target-feature -zksed -target-feature -zksh -target-feature -zkt -target-feature -zmmul -target-feature -zvbb -target-feature -zvbc -target-feature -zve32f -target-feature -zve32x -target-feature -zve64d -target-feature -zve64f -target-feature -zve64x -target-feature -zvfh -target-feature -zvfhmin -target-feature -zvkb -target-feature -zvkg -target-feature -zvkn -target-feature -zvknc -target-feature -zvkned -target-feature -zvkng -target-feature -zvknha -target-feature -zvknhb -target-feature -zvks -target-feature -zvksc -target-feature -zvksed -target-feature -zvksg -target-feature -zvksh -target-feature -zvkt -target-feature -zvl1024b -target-feature -zvl128b -target-feature -zvl16384b -target-feature -zvl2048b -target-feature -zvl256b -target-feature -zvl32768b -target-feature -zvl32b -target-feature -zvl4096b -target-feature -zvl512b -target-feature -zvl64b -target-feature -zvl65536b -target-feature -zvl8192b -target-feature -experimental-zacas -target-feature -experimental-zcmop -target-feature -experimental-zfbfmin -target-feature -experimental-zicfilp -target-feature -experimental-zicfiss -target-feature -experimental-zicond -target-feature -experimental-zimop -target-feature -experimental-ztso -target-feature -experimental-zvfbfmin -target-feature -experimental-zvfbfwma -target-feature +relax -target-abi lp64d -msmall-data-limit 8 -debugger-tuning=gdb -fdebug-compilation-dir=/scratch/tc-testing/llvm-fuzz-ci/triage-7-4127 -fcoverage-compilation-dir=/scratch/tc-testing/llvm-fuzz-ci/triage-7-4127 -resource-dir /scratch/tc-testing/tc-jan-17-llvm/build/lib/clang/18 -isysroot /scratch/tc-testing/tc-jan-17-llvm/build/bin/../sysroot -internal-isystem /scratch/tc-testing/tc-jan-17-llvm/build/lib/clang/18/include -internal-isystem /scratch/tc-testing/tc-jan-17-llvm/build/bin/../sysroot/usr/local/include -internal-isystem /scratch/tc-testing/tc-jan-17-llvm/build/bin/../sysroot/lib/gcc/riscv64-unknown-linux-gnu/12.2.0/../../../../riscv64-unknown-linux-gnu/include -internal-externc-isystem /scratch/tc-testing/tc-jan-17-llvm/build/bin/../sysroot/include -internal-externc-isystem /scratch/tc-testing/tc-jan-17-llvm/build/bin/../sysroot/usr/include -O3 -ferror-limit 19 -fno-signed-char -fgnuc-version=4.2.1 -fcolor-diagnostics -vectorize-loops -vectorize-slp -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /scratch/tmp/red-a806a3.o -x c red.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'red.c'.
4.      Running pass 'RISC-V Merge Base Offset' on function '@main'
 #0 0x0000558363cb3810 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x1ba8810)
 #1 0x0000558363cb0dde SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f59cea42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #3 0x0000558362e7189b (anonymous namespace)::RISCVMergeBaseOffsetOpt::runOnMachineFunction(llvm::MachineFunction&) RISCVMergeBaseOffset.cpp:0:0
 #4 0x00005583632103b0 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
 #5 0x000055836378e67a llvm::FPPassManager::runOnFunction(llvm::Function&) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x168367a)
 #6 0x000055836378e804 llvm::FPPassManager::runOnModule(llvm::Module&) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x1683804)
 #7 0x000055836378f244 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x1684244)
 #8 0x0000558363ef8502 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x1ded502)
 #9 0x000055836449fb91 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x2394b91)
#10 0x0000558365ddc009 clang::ParseAST(clang::Sema&, bool, bool) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x3cd1009)
#11 0x00005583647317f1 clang::FrontendAction::Execute() (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x26267f1)
#12 0x00005583646b04b9 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x25a54b9)
#13 0x00005583647f3ad3 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0x26e8ad3)
#14 0x0000558362d21503 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0xc16503)
#15 0x0000558362d19bed ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#16 0x0000558362d1d78a clang_main(int, char**, llvm::ToolContext const&) (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0xc1278a)
#17 0x0000558362c683db main (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0xb5d3db)
#18 0x00007f59cea29d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#19 0x00007f59cea29e40 call_init ./csu/../csu/libc-start.c:128:20
#20 0x00007f59cea29e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#21 0x0000558362d19065 _start (/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang-18+0xc0e065)
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 18.0.0git (https://github.com/llvm/llvm-project.git 430a40d12eaa5a61792c4670955c110146902afb)
Target: riscv64-unknown-linux-gnu
Thread model: posix
InstalledDir: /scratch/tc-testing/tc-jan-17-llvm/build/bin
clang: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /scratch/tmp/red-796f4b.c
clang: note: diagnostic msg: /scratch/tmp/red-796f4b.sh
clang: note: diagnostic msg:

********************
  ```
</details>

[tmp.zip](https://github.com/llvm/llvm-project/files/13985885/tmp.zip)

```
/scratch/tc-testing/tc-jan-17-llvm/build/bin/clang red.c -O3 -emit-llvm -c -o foo.bc
/scratch/tc-testing/tc-jan-17-llvm/build/build-llvm-linux/bin/llc foo.bc -relocation-model=pic
/scratch/tc-testing/tc-jan-17-llvm/build/build-llvm-linux/bin/bugpoint -run-llc foo.bc --tool-args -relocation-model=pic
```
[bugpoint-reduced-simplified.zip](https://github.com/llvm/llvm-project/files/13985980/bugpoint-reduced-simplified.zip)

This is my first LLVM bug so please let me know if I'm missing anything :smile: 

This bug was found with csmith which is a c-level fuzzer that I've been using on GCC for a while now. If you're interested or want more info here's the repo (Still in the process of adapting my scripts for LLVM fuzzing/reducing):
https://github.com/patrick-rivos/gcc-fuzz-ci

### PR DESCRIPTION
**Pull request by topperc**
Fixes #78679.

### CODE REVIEW

**Comment by jrtc27** [comment]
```MachineInstr &OffsetTail = *MRI->getVRegDef(Reg);``` also appears twice in the file, don't those need guarding? Maybe hasOneUse happens to be sufficient somehow to prevent that, but if so it's not immediately obvious to me why.

**Comment by topperc** [comment]
> `MachineInstr &OffsetTail = *MRI->getVRegDef(Reg);` also appears twice in the file, don't those need guarding? Maybe hasOneUse happens to be sufficient somehow to prevent that, but if so it's not immediately obvious to me why.

They seem pretty unlikely. It would mean we have a ADD or SHXADD with an X0 or other physical register operand. ISel doesn't create those today and we haven't ran the coalescer yet so I don't think it can happen. But I'll fix them too.

File: llvm/test/CodeGen/RISCV/fold-addi-loadstore.ll [hunk 1]
```diff
@@ -964,3 +964,69 @@ for.body:                                         ; preds = %for.body.lr.ph, %fo
 }
 
 declare void @f(ptr)
+
+@g = external dso_local global [100 x [100 x i8]]
+
+; This Test used to crash due to caling getVRegDef on X0.
```
Comment (asb) [inline llvm/test/CodeGen/RISCV/fold-addi-loadstore.ll:970]: Nit: Test => test, and caling => calling
(Sorry!)

**Review by asb — APPROVED** [review]
LGTM.

**Pull request merged**